visualization/app/codeCharta/core/data/data.deltaCalculator.service.ts

```diff
@@ -48,10 +49,120 @@ export class DeltaCalculatorService {
 
     }
 
-    calculateAttributeListDelta(first: KVObject, second: KVObject){
+    /**
+     * Unifies both maps and inserts unique nodes into the other map. The inserted nodes have all metrics set to zero.
+     * This will calculate deltas correctly between versions and add "empty" buildings to ensure all revisions have an
+     * equal building pool. These empty buildings should be viewed with the unary metric as area
+     * @param {CodeMap} leftMap
+     * @param {CodeMap} rightMap
+     * @returns {{leftMap: CodeMap; rightMap: CodeMap}}
+     */
+    public fillMapsWithNonExistingNodesFromOtherMap(leftMap: CodeMap, rightMap: CodeMap): { leftMap: CodeMap; rightMap: CodeMap; } {
+        if (leftMap && rightMap && leftMap.root && rightMap.root) {
+
+            let leftMapCopy: CodeMap = deepcopy.default(leftMap);
+            let rightMapCopy: CodeMap = deepcopy.default(rightMap);
+
+            let leftRoot = d3.hierarchy<CodeMapNode>(leftMapCopy.root);
+            let rightRoot = d3.hierarchy<CodeMapNode>(rightMapCopy.root);
+
+            this.insertLeftIntoRightWithZeroMetrics(leftRoot, rightRoot);
+            this.insertLeftIntoRightWithZeroMetrics(rightRoot, leftRoot);
+
+            return {leftMap: leftMapCopy, rightMap: rightMapCopy};
+
+        } else {
+            return {leftMap: leftMap, rightMap: rightMap};
+        }
+    }
+
+    public removeUpCrossOriginNodes(map: CodeMap): CodeMap {
+        if (map && map.root) {
+
+            let mapCopy: CodeMap = deepcopy.default(map);
+
+            let mapRoot = d3.hierarchy<CodeMapNode>(mapCopy.root);
+            mapRoot.each((node)=>{
+                if(node.data.children) {
+                    node.data.children = node.data.children.filter(x => (x.origin === mapCopy.fileName));
+                }
+            });
+
+            return mapCopy;
+
+        } else {
+            return map;
+        }
+    }
+
+    private insertLeftIntoRightWithZeroMetrics(left: HierarchyNode<CodeMapNode>, right: HierarchyNode<CodeMapNode>) {
```

Comment:
may factor in function in smaller subfunctions, corresponding to the code comments?